design/ironic-standalone-networking-service.md

```diff
@@ -0,0 +1,734 @@
+  capabilities provided by networking-generic-switch
+- Implementing custom switch drivers (users will use existing drivers from
+  networking-generic-switch)
+- Managing switch infrastructure itself (firmware updates, switch provisioning,
+  etc.)
+- Supporting dynamic network reconfiguration after node deployment (initial
+  scope is provisioning-time configuration only)
+
+## Proposal
+
+This proposal introduces four main components:
+
+1. **New CRDs**: `BareMetalSwitch` and `BareMetalNetworkAttachment` to define
+   switch configurations and network interface configurations respectively
+2. **Enhanced Provision CRD**: Add `enableIronicNetworking` flag and service
+   network configuration to the `provisions.metal3.io` CRD
+3. **Enhanced BareMetalHost CRD**: Add `networkInterfaces` field to reference
+   network configurations for each network interface
+4. **Ironic Networking container**: Deploy the Ironic Networking service
+   alongside the existing metal3-ironic container
+
+### User Stories
+
+#### Story 1
+
+As a Metal³ operator, I want to define my ToR switch configurations once in
+Kubernetes CRDs so that Metal³ can automatically manage switch port
+configurations without manual intervention.
+
+The operator creates `BareMetalSwitch` resources for each ToR switch:
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalSwitch
+metadata:
+  name: rack1-tor1
+spec:
+  driverType: generic-switch
+  deviceType: dell_os10
+  address: 192.168.1.10
+  credentials:
+    type: "password"
+    secretName: switch-credentials
+  macAddress: 3e:c5:c0:cb:fa:f6
+```
+
+Metal³ automatically generates the Ironic Networking switch configuration file
+and restarts the service when switches are added or modified.
+
+The credentials Secret is expected to contain keys for the `username`,
+`password`, and (optionally) `enable-secret`.
+
+#### Story 2
+
+As a Metal³ user, I want to define network configurations for my bare metal
+hosts declaratively so that switch ports are automatically configured when nodes
+are provisioned.
+
+The user creates a `BareMetalNetworkAttachment` defining the desired network
+configuration for an access mode port for the primary network interface:
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalNetworkAttachment
+metadata:
+  name: machine-network-vlan100
+spec:
+  mode: access
+  nativeVlan: 100
+```
+
+The user optionally creates other `BareMetalNetworkAttachment` instances
+defining the desired network configuration for trunk ports for the secondary
+network interface(s):
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalNetworkAttachment
+metadata:
+  name: data-network-vlans-200-202
+spec:
+  mode: trunk
+  allowedVlans: [200, 201, 202]
+```
+
+Then references them in the `BareMetalHost`:
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: worker-1
+spec:
+  # ... other fields ...
+  networkInterfaces:
+  - macAddress: "aa:bb:cc:dd:ee:01"
```

Comment:
Is there any value in making each record here a separate CRD (like HostNetworkInterface)? This is not a suggestion, I'm genuinely asking.